includes/class-micropub-authorize.php

```diff
@@ -0,0 +1,218 @@
+		return static::$micropub_auth_response;
+	}
+
+	public static function header( $header, $value ) {
+			header( $header . ': ' . $value, false );
+	}
+
+	public static function http_header() {
+			static::header( 'Link', '<' . get_option( 'indieauth_authorization_endpoint', MICROPUB_AUTHENTICATION_ENDPOINT ) . '>; rel="authorization_endpoint"' );
+			static::header( 'Link', '<' . get_option( 'indieauth_token_endpoint', MICROPUB_TOKEN_ENDPOINT ) . '>; rel="token_endpoint"' );
+	}
+	public static function html_header() {
+			printf( '<link rel="authorization_endpoint" href="%s" />' . PHP_EOL, get_option( 'indieauth_authorization_endpoint', MICROPUB_AUTHENTICATION_ENDPOINT ) );
+			printf( '<link rel="token_endpoint" href="%s" />' . PHP_EOL, get_option( 'indieauth_token_endpoint', MICROPUB_TOKEN_ENDPOINT ) );
+	}
+
+	public static function jrd_links( $array ) {
+		$array['links'][] = array(
+			'rel'  => 'authorization_endpoint',
+			'href' => get_option( 'indieauth_authorization_endpoint', MICROPUB_AUTHENTICATION_ENDPOINT ),
+		);
+		$array['links'][] = array(
+			'rel'  => 'token_endpoint',
+			'href' => get_option( 'indieauth_token_endpoint', MICROPUB_TOKEN_ENDPOINT ),
+		);
+
+		return $array;
+	}
+
+	public static function get( $array, $key, $default = array() ) {
+		if ( is_array( $array ) ) {
+			return isset( $array[ $key ] ) ? $array[ $key ] : $default;
+		}
+		return $default;
+	}
+
+	/**
+	 * Get the authorization header
+	 *
+	 * On certain systems and configurations, the Authorization header will be
+	 * stripped out by the server or PHP. Typically this is then used to
+	 * generate `PHP_AUTH_USER`/`PHP_AUTH_PASS` but not passed on. We use
+	 * `getallheaders` here to try and grab it out instead.
+	 *
+	 * @return string|null Authorization header if set, null otherwise
+	 */
+	public static function get_authorization_header() {
+		if ( ! empty( $_SERVER['HTTP_AUTHORIZATION'] ) ) {
+			return wp_unslash( $_SERVER['HTTP_AUTHORIZATION'] );
+		}
+		// When Apache speaks via FastCGI with PHP, then the authorization header is often available as REDIRECT_HTTP_AUTHORIZATION.
+		if ( ! empty( $_SERVER['REDIRECT_HTTP_AUTHORIZATION'] ) ) {
+			return wp_unslash( $_SERVER['REDIRECT_HTTP_AUTHORIZATION'] );
+		}
+		$headers = getallheaders();
+		// Check for the authorization header case-insensitively
+		foreach ( $headers as $key => $value ) {
+			if ( strtolower( $key ) === 'authorization' ) {
+				return $value;
+			}
+		}
+		return null;
+	}
+
+	/**
+	 * Validate the access token at the token endpoint.
+	 *
+	 * https://indieauth.spec.indieweb.org/#access-token-verification
+	 */
+	public static function determine_current_user( $user_id ) {
+		$cls = get_called_class();
+		// Do not try to find a user if one has already been found
+		if ( ! empty( $user_id ) ) {
+			return $user_id;
+		}
+
+		// Only try to authenticate if this is the Micropub endpoint
+		if ( ! empty( get_query_var( 'micropub' ) ) ) {
```

Comment:
very minor nit: for readability, consider adding something like:

> ..., since this runs as a filter on the built in `determine_current_user` function.

as is, it took me a few minutes to figure out why we had this. 😁